packages/material-ui-lab/src/BusyButton/BusyButton.js

```diff
@@ -0,0 +1,144 @@
+    visibility: 'visible',
+    display: 'flex',
+  },
+  /* Styles applied to the pendingIndicator element if `pendingPosition="center"`. */
+  pendingIndicatorCenter: {
+    left: '50%',
+    transform: 'translate(-50%)',
+  },
+  /* Styles applied to the pendingIndicator element if `pendingPosition="start"`. */
+  pendingIndicatorStart: {
+    left: 14,
+  },
+  /* Styles applied to the pendingIndicator element if `pendingPosition="end"`. */
+  pendingIndicatorEnd: {
+    right: 14,
+  },
+  /* Styles applied to the endIcon element if `pending={true}` and `pendingPosition="end"`. */
+  endIconLoadingEnd: {
+    visibility: 'hidden',
+  },
+  /* Styles applied to the startIcon element if `pending={true}` and `pendingPosition="start"`. */
+  startIconLoadingStart: {
+    visibility: 'hidden',
+  },
+  /* Styles applied to the label element if `pending={true}` and `pendingPosition="center"`. */
+  labelLoadingCenter: {
+    visibility: 'hidden',
+  },
+});
+
+const PendingIndicator = <CircularProgress color="inherit" size={16} />;
+
+const BusyButton = React.forwardRef(function BusyButton(props, ref) {
+  const {
+    children,
+    classes,
+    className,
+    disabled = false,
+    pending = false,
+    pendingIndicator = PendingIndicator,
+    pendingPosition = 'center',
+    ...other
+  } = props;
+
+  return (
+    <Button
+      className={clsx(
+        classes.root,
+        {
+          [classes.pending]: pending,
+        },
+        className,
+      )}
+      disabled={disabled || pending}
+      ref={ref}
+      classes={{
+        startIcon: classes[`startIcon${pending ? 'Loading' : ''}${capitalize(pendingPosition)}`],
+        endIcon: classes[`endIcon${pending ? 'Loading' : ''}${capitalize(pendingPosition)}`],
+        label: classes[`label${pending ? 'Loading' : ''}${capitalize(pendingPosition)}`],
+      }}
+      {...other}
+    >
+      {pending && (
+        <div
+          className={clsx(
+            classes.pendingIndicator,
+            classes[`pendingIndicator${capitalize(pendingPosition)}`],
+          )}
+        >
+          {pendingIndicator}
+        </div>
+      )}
+
+      {children}
+    </Button>
+  );
+});
+
+BusyButton.propTypes = {
+  // ----------------------------- Warning --------------------------------
+  // | These PropTypes are generated from the TypeScript type definitions |
+  // |     To update them edit the d.ts file and run "yarn proptypes"     |
+  // ----------------------------------------------------------------------
+  /**
+   * The content of the button.
+   */
+  children: PropTypes.node,
+  /**
+   * Override or extend the styles applied to the component.
+   * See [CSS API](#css) below for more details.
+   */
+  classes: PropTypes.object,
+  /**
+   * @ignore
+   */
+  className: PropTypes.string,
+  /**
+   * If `true`, the button will be disabled.
+   */
+  disabled: PropTypes.bool,
+  /**
+   * The content of the button.
+   */
+  pending: PropTypes.bool,
```

Comment:
Sorry, I know this was already discussed somewhere in all the resolved comments, but what do you think about `busy`, `busyIndicator`, `busyPosition` for these props given the BusyButton component name/

```
<BusyButton busy busyIndicator="Loading...">Load</BusyButton>

Reply:
Here is the thread with the decision on the naming https://github.com/mui-org/material-ui/pull/21389#discussion_r438994929 :) 

In short words, we were trying to follow React's naming for similar use-cases.

Reply:
I'm still not satisfied with the name. It's peek-bikeshedding but it just irks me that we have a "busy button" that isn't "busy" e.g. `<BusyButton busy={false} />` but you rarely hardcode it anyway so :man_shrugging: 

I only feel strongly about the prop naming. Using the same name as you'd use with Suspense features makes it obvious what belongs together. "TransitionButton" would be my favorite but "transition" is too generic I fear.

Reply:
Thanks @mnajdova, that makes sense!

Reply:
@eps1lon No strong preference either way about the component name, but LoadingButton does seem [quite common](https://google.com/search?q=loading+button), even if it isn't technically correct.

Reply:
@mbrookes I had a similar thought: 

<img width="340" alt="Capture d’écran 2020-06-14 à 23 51 25" src="https://user-images.githubusercontent.com/3165635/84604976-ff9e1180-ae99-11ea-9c1d-85107b06ac2c.png">

Reply:
At this point I don’t remember why I’ve rename it in the end. Should we go back with LoadingButton?

Reply:
I've renamed it to `LoadingButton`, that's the name everyone uses so 🤷‍♂️

Reply:
> At this point I don’t remember why I’ve rename it in the end.

Olivier's fault :stuck_out_tongue: :  https://github.com/mui-org/material-ui/pull/21389#issuecomment-642856473